### PR TITLE
upstream: add transport socket failure reason to stream info and log

### LINKED ISSUE #5603: access log: log TLS related errors for HTTP
*Description*:

While upstream TLS is configured, HTTP access log is not aware of any of them. [Access log response code](https://www.envoyproxy.io/docs/envoy/latest/configuration/access_log) UF, UR, UC are potential code for TLS errors such as peer certificate expired, not trusted, protocol or cipher mismatch handshake timeout etc. It would be nice if we could do for downstream too but we don't have enough request information, though for upstream the log should be aware.

### PR DESCRIPTION
**Pull request by lizan**
*Description*:
Fixes #5603 

*Risk Level*: Low (not changing flow, add more info)
*Testing*: unit test,
*Docs Changes*: Added
*Release Notes*: Added

### CODE REVIEW

File: docs/root/configuration/access_log.rst [hunk 1]
```diff
@@ -228,6 +228,13 @@ The following command operators are supported:
   Local address of the upstream connection. If the address is an IP address it includes both
   address and port.
 
+%UPSTREAM_TRANSPORT_FAILURE_REASON%
```
Comment (mattklein123) [inline docs/root/configuration/access_log.rst]: Can we have parity here with the gRPC access log? I would like to make sure we have consistency with all of these types of changes moving forward.

File: docs/root/configuration/access_log.rst [hunk 2]
```diff
@@ -228,6 +228,13 @@ The following command operators are supported:
...
+  HTTP
+    The reason of upstream transport failure, this is usually upstream TLS connection failure with 503 response code.
```
Comment (mattklein123) [inline docs/root/configuration/access_log.rst]: Can you make this a little more clear? Is the latter part of the sentence the actual error message? Should it be in quotes? Also, do we want a string here, or should we have well defined error codes like we do for other failures?

File: docs/root/configuration/access_log.rst [hunk 3]
```diff
@@ -228,6 +228,13 @@ The following command operators are supported:
...
+
+  TCP
+    Not implemented ("-")
```
Comment (mattklein123) [inline docs/root/configuration/access_log.rst]: We could implement this, right? Is there a TODO somewhere?

File: docs/root/configuration/access_log.rst (hunk 2 above)
Comment (lizan) [reply docs/root/configuration/access_log.rst]: I considered well defined error codes but I chose a string here due to the transport socket extensibility and it is likely only used in logging purpose. We can make some well known errors as predefined strings though. No the latter part is not the actual message, will update the document.

File: docs/root/configuration/access_log.rst (hunk 1 above)
Comment (lizan) [reply docs/root/configuration/access_log.rst]: Done

File: docs/root/configuration/access_log.rst (hunk 3 above)
Comment (lizan) [reply docs/root/configuration/access_log.rst]: Added a TODO, will look into this later to see if I can implement within this PR.

File: source/common/router/router.h [hunk 4]
```diff
@@ -380,7 +381,8 @@ class Filter : Logger::Loggable<Logger::Id::router>,
   void onUpstreamMetadata(Http::MetadataMapPtr&& metadata_map);
   void onUpstreamComplete();
   void onUpstreamReset(UpstreamResetType type,
-                       const absl::optional<Http::StreamResetReason> reset_reason);
+                         const absl::optional<Http::StreamResetReason> &reset_reason,
```
Comment (mattklein123) [inline source/common/router/router.h]: nit: no reason to pass by reference

File: source/extensions/transport_sockets/tls/ssl_socket.cc [hunk 5]
```diff
@@ -21,12 +21,16 @@ namespace TransportSockets {
 namespace Tls {
 
 namespace {
+
+constexpr absl::string_view NotReadyReason{"SSL error: secret is not ready"};
```
Comment (mattklein123) [inline source/extensions/transport_sockets/tls/ssl_socket.cc]: nit: What does this mean? Can you make the error message a bit better or potentially actually document all the potential errors for the built-in transport sockets somewhere?

File: test/extensions/transport_sockets/tls/ssl_socket_test.cc [hunk 6]
```diff
@@ -599,6 +599,14 @@ const std::string testUtilV2(const TestUtilOptionsV2& options) {
     EXPECT_EQ(1UL, client_stats_store.counter(options.expectedClientStats()).value());
   }
 
+  if (options.expectSuccess()) {
+    EXPECT_EQ("", client_connection->transportFailureReason());
+    EXPECT_EQ("", server_connection->transportFailureReason());
+  } else {
+    EXPECT_NE("", client_connection->transportFailureReason());
```
Comment (mattklein123) [inline test/extensions/transport_sockets/tls/ssl_socket_test.cc]: Should we test a specific failure reason here?

Comment (lizan) [reply test/extensions/transport_sockets/tls/ssl_socket_test.cc]: Done

File: source/extensions/transport_sockets/tls/ssl_socket.cc (hunk 5 above)
Comment (lizan) [reply source/extensions/transport_sockets/tls/ssl_socket.cc]: I hope now it is better, when SDS is configured there are chances that Cluster are up but Secret is not ready.

File: source/common/router/router.h (hunk 4 above)
Comment (lizan) [reply source/common/router/router.h]: ah, bad merge, fixed. btw this shouldn't be const if passing by value though.

**Comment by lizan** [comment]
> I think I would prefer if we had some more documentation for the known messages that the built-in transport sockets emit. Can we add that?

Done.

> Also, are there any relevant errors we should print for the non-TLS transport? I guess I'm wondering if we should do the equivalent of strerror to print out connection failure reasons, etc.? Might not be worth it in this PR but maybe a TODO somewhere if you think that would be relevant?

Yes that's one of the reason I put it into Connection interface, left a TODO there.

File: docs/root/intro/arch_overview/ssl.rst [hunk 7]
```diff
@@ -158,3 +158,22 @@ infrastructure.
 
 Client TLS authentication filter :ref:`configuration reference
 <config_network_filters_client_ssl_auth>`.
+
+.. _arch_overview_ssl_trouble_shooting:
+
+Trouble shooting
+----------------
+
+When Envoy originates TLS when making connections to upstream clusters, its error will be logged into
```
Comment (mattklein123) [inline docs/root/intro/arch_overview/ssl.rst]: nit: "..., any errors will be logged into the ..."

File: docs/root/intro/arch_overview/ssl.rst [hunk 8]
```diff
@@ -158,3 +158,22 @@ infrastructure.
...
+:ref:`UPSTREAM_TRANSPORT_FAILURE_REASON<config_access_log_format_upstream_transport_failure_reason>` field.
+Common errors are:
+
+* ``Secret is not delivered via SDS``: Envoy is still waiting SDS to deliver key/cert or root CA.
```
Comment (mattklein123) [inline docs/root/intro/arch_overview/ssl.rst]: nit: I think I would call this error "Secret not supplied by SDS" WDYT?
nit: "still waiting for"

File: docs/root/intro/version_history.rst [hunk 9]
```diff
@@ -6,6 +6,7 @@ Version history
 * access log: added a new flag for upstream retry count exceeded.
 * access log: added a :ref:`gRPC filter <envoy_api_msg_config.filter.accesslog.v2.GrpcStatusFilter>` to allow filtering on gRPC status.
 * access log: added a new flag for stream idle timeout.
+* access log: added UPSTREAM_TRANSPORT_FAILURE_REASON for HTTP.
```
Comment (mattklein123) [inline docs/root/intro/version_history.rst]: nit: can you ref link to both new proto field and text formatting docs?

**Comment by lizan** [comment]
@mattklein123 docs updated, let's merge #6142 first and I will update this PR. I believe they conflicts.

File: source/extensions/transport_sockets/tls/ssl_socket.cc [hunk 10]
```diff
@@ -22,7 +22,7 @@ namespace Tls {
 
 namespace {
 
-constexpr absl::string_view NotReadyReason{"TLS error: Secret is not delivered via SDS"};
+constexpr absl::string_view NotReadyReason{"TLS error: Secret is not supplied by SDS"};
```
Comment (mattklein123) [inline source/extensions/transport_sockets/tls/ssl_socket.cc]: Would be nice if there was a test that covers this?

Comment (lizan) [reply source/extensions/transport_sockets/tls/ssl_socket.cc]: done

**Pull request merged**